File: exercism/ruby/palindrome-products/palindrome_products.rb
class Integer
  def palindrome?
    as_string = self.to_s
    as_string == as_string.reverse
  end
end

class Palindromes
  def initialize(max_factor:, min_factor: 1)
    @min, @max = min_factor, max_factor
  end
  def generate
    @pals = (@min .. @max).to_a.repeated_combination(2)
                          .select{|i, j| (i * j).palindrome? }
                          .group_by{|i, j| i * j }
  end
  def largest
    Palindrome.new(*@pals.max_by{|k, _| k })
  end
  def smallest
    Palindrome.new(*@pals.min_by{|k, _| k })
  end
end

class Palindrome
  attr_reader :value, :factors
  def initialize(v, fs)
    @value = v
    @factors = fs
  end
end

File: exercism/elixir/palindrome-products/palindrome_products.exs
defmodule Palindromes do

  @doc """
  Generates all palindrome products from an optionally given min factor (or 1) to a given max factor.
  """
  @spec generate(non_neg_integer, non_neg_integer) :: map() 
  def generate(max_factor, min_factor \\ 1) do
    (for a <- min_factor..max_factor,
         b <- min_factor..max_factor,
         a <= b,
         is_palindrome?(a*b),
         do: [a, b])
      |> Enum.reverse
      |> Enum.reduce(Map.new, &insert/2)
  end

  defp insert([a, b], dict) do
    Map.update(dict, a*b, [[a, b]], &([[a, b] | &1]))
  end
  defp is_palindrome?(number) do
    to_char_list(number) == Enum.reverse(to_char_list(number))
  end
end
